esphome/components/sml/text_sensor/sml_text_sensor.cpp

```diff
@@ -0,0 +1,55 @@
+#include "esphome/core/helpers.h"
+#include "esphome/core/log.h"
+#include "sml_text_sensor.h"
+#include "../sml_parser.h"
+
+namespace esphome {
+namespace sml {
+
+static const char *const TAG = "sml_text_sensor";
+
+SmlTextSensor::SmlTextSensor(std::string server_id, std::string obis, std::string format)
+    : SmlListener(std::move(server_id), std::move(obis)), format_(std::move(format)) {}
+
+void SmlTextSensor::publish_val(const ObisInfo &obis_info) {
+  char value_type;
+  if (this->format_ == "hex")
+    value_type = SML_HEX;
+  else if (this->format_ == "text")
+    value_type = SML_OCTET;
+  else if (this->format_ == "bool")
+    value_type = SML_BOOL;
+  else if (this->format_ == "uint")
+    value_type = SML_UINT;
+  else if (this->format_ == "int")
+    value_type = SML_INT;
+  else
+    value_type = obis_info.value_type;
+
+  switch (value_type) {
+    case SML_HEX: {
+      publish_state("0x" + bytes_repr(obis_info.value));
+      break;
+    }
+    case SML_INT: {
+      publish_state(to_string(bytes_to_int(obis_info.value)));
+      break;
+    }
+    case SML_BOOL:
+      publish_state(bytes_to_uint(obis_info.value) ? "True" : "False");
+      break;
+    case SML_UINT: {
+      publish_state(to_string(bytes_to_uint(obis_info.value)));
+      break;
+    }
+    case SML_OCTET: {
+      publish_state(std::string(obis_info.value.begin(), obis_info.value.end()));
```

Comment:
```suggestion
      publish_state(std::string(obis_info.value.begin(), obis_info.value.end()));
```

`text_sensor` can effectively only represent UTF-8 data (MQTT, native API clients both transport it that way and would throw errors if non-UTF8 is received)

Does OBIS specify that the incoming data must be UTF-8/ASCII?

Reply:
The protocol specifications do not say anything about encodings. The only "datatype" given is a byte sequence (which they call an "octet string").
What data is tranported and how it should be interpreted differs from manufacturer and from device to device.
Unfortunatley it's up to the user to find the right interpretation.

Reply:
:+1: 

If there _is_ non-UTF-8 data sent by the device, the user can still select `HEX` output mode and it will just print the hex encoding of the byte sequence, right? (just so we have a fallback if someone does hit that as an error)

Reply:
Yes, that's the way it works. If necessary, the user can do whatever he wants with the hex representation using templates on the HomeAssistant side.